cmd/disk-cache-backend.go

```diff
@@ -283,49 +305,121 @@ func (c *diskCache) Stat(ctx context.Context, bucket, object string) (oi ObjectI
 	return
 }
 
-// statCache is a convenience function for purge() to get ObjectInfo for cached object
-func (c *diskCache) statCache(cacheObjPath string) (oi ObjectInfo, e error) {
+// statCachedMeta returns metadata from cache - including ranges cached, partial to indicate
+// if partial object is cached.
+func (c *diskCache) statCachedMeta(cacheObjPath string) (meta *cacheMeta, partial bool, err error) {
 	// Stat the file to get file size.
 	metaPath := path.Join(cacheObjPath, cacheMetaJSONFile)
 	f, err := os.Open(metaPath)
 	if err != nil {
-		return oi, err
+		return meta, partial, err
 	}
 	defer f.Close()
-
-	meta := &cacheMeta{Version: cacheMetaVersion}
+	meta = &cacheMeta{Version: cacheMetaVersion}
 	if err := jsonLoad(f, meta); err != nil {
-		return oi, err
+		return meta, partial, err
 	}
+	// get metadata of part.1 if full file has been cached.
+	partial = true
 	fi, err := os.Stat(pathJoin(cacheObjPath, cacheDataFile))
+	if err == nil {
+		meta.Stat.ModTime = atime.Get(fi)
+		partial = false
+	}
+	return meta, partial, nil
+}
+
+// statRange returns ObjectInfo and rangeInfo from disk cache
+func (c *diskCache) statRange(ctx context.Context, bucket, object string, rs *HTTPRangeSpec) (oi ObjectInfo, rngInfo rangeInfo, err error) {
+	// Stat the file to get file size.
+	cacheObjPath := getCacheSHADir(c.dir, bucket, object)
+
+	if rs == nil {
+		oi, err = c.statCache(cacheObjPath)
+		return oi, rngInfo, err
+	}
+	meta, _, err := c.statCachedMeta(cacheObjPath)
+	if err != nil {
+		return oi, rngInfo, err
+	}
+
+	actualSize := uint64(meta.Stat.Size)
+
+	off, length, err := rs.GetOffsetLength(int64(actualSize))
+	if err != nil {
+		return oi, rngInfo, err
+	}
+	actualRngSize := uint64(length)
+	if globalCacheKMS != nil {
+		actualRngSize, _ = sio.EncryptedSize(uint64(length))
+	}
+	rng := fmt.Sprintf("%d-%d", off, off+length-1)
```

Comment:
can you write `rs.String()` version to write a stringified version of this? - looks repetitive to do this everywhere.